Framework/DataObjects/src/TimeSplitter.cpp

```diff
@@ -280,5 +293,171 @@ TimeROI TimeSplitter::getTimeROI(const int workspaceIndex) {
+ */
+void TimeSplitter::splitEventList(const EventList &events, std::map<int, EventList *> partials, bool pulseTof,
+                                  bool tofCorrect, double factor, double shift) const {
+
+  if (events.getEventType() == EventType::WEIGHTED_NOTIME)
+    throw std::runtime_error("EventList::splitByTime() called on an EventList "
+                             "that no longer has time information.");
+
+  this->initializePartials(events, partials);
+
+  if (this->empty())
+    return;
+
+  // sort the list of events and return their times, in nanoseconds
+  std::vector<int64_t> times = this->sortEventList(events, pulseTof, tofCorrect, factor, shift);
+
+  switch (events.getEventType()) {
+  case EventType::TOF:
+    this->splitEventVec(times, events.getEvents(), partials);
+    break;
+  case EventType::WEIGHTED:
+    this->splitEventVec(times, events.getWeightedEvents(), partials);
+    break;
+  default:
+    throw std::runtime_error("Unhandled event type");
+  }
+}
+
+/**
+ * Initialize the detector ID's and event type of the destination event lists.
+ * @param events : list of input events
+ * @param partials : resulting partial lists of events
+ */
+void TimeSplitter::initializePartials(const EventList &events, std::map<int, EventList *> partials) const {
+  // collect the state from events which is to be transferred to the partials
+
+  bool removeDetIDs{true};
+  auto detIDs = events.getDetectorIDs();
+  auto histogram = events.getHistogram();
+  auto eventType = events.getEventType();
+  // lambda expression initializing one partial
+  auto initPartial = [&](EventList *partial) {
+    partial->clear(removeDetIDs);
+    partial->setDetectorIDs(detIDs);
+    partial->setHistogram(histogram); // problem here!
+    partial->switchTo(eventType);
+  };
+  // iterate over the partials
+  for (auto iter = cbegin(partials); iter != cend(partials); ++iter)
+    initPartial(iter->second);
+}
+
+/**
+ * In-place sort a list of input events either by Pulse time or by Pulse+TOF time.
+ *
+ * The returned sorted vector of times will contain one of the following list of times:
+ * - Pulse time for each event, when pulseTof==false
+ * - Pulse + TOF time for each event, when pulseTof==true && tofCorrect==false
+ * - Pulse + (factor * TOF + shift) for each event, when pulseTof==true && tofCorrect==true
+ *
+ * @param events : list of input events
+ * @param pulseTof : if True, split according to Pulse + TOF time, otherwise split by Pulse time
+ * @param tofCorrect : rescale and shift the TOF values (factor*TOF + shift)
+ * @param factor : rescale the TOF values by a dimensionless factor.
+ * @param shift : shift the TOF values after rescaling, in units of microseconds.
+ * @return : times that will be compared again the splitter boundaries, in units of nanoseconds.
+ */
+std::vector<int64_t> TimeSplitter::sortEventList(const EventList &events, bool pulseTof, bool tofCorrect, double factor,
```

Comment:
I think this method could be greatly simplified by using `EventList::sort*` methods (picking the right one) and `EventList::getPulseTimes()`. I think this is correct in assuming that `EventList::sortTimeAtSample()` will give the same result as `EventList::sortPulseTime()` when `factor>=0`. The missing bit to make this "simpler" version work is an extra method `EventList::getTimesAtSample(const double &tofFactor, const double &tofOffset)` which would still return a `std::vector<DateAndTime>`.